### ISSUE TITLE
Pulling an nonexistent image returns 200 instead of 404

### ISSUE THREAD

**Issue by fermayo** [moby/moby#32102]
**Description**
When pulling an image from a `docker/distribution` registry that does not exist, the `/images/create` API gives an HTTP response code of 200 instead of 404 as per the documentation.

**Steps to reproduce the issue:**
1. Launch a registry in localhost
2. Run `curl --unix-socket /var/run/docker.sock -X POST http://localhost/images/create\?fromImage\=127.0.0.1:5000/docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8 -v`

**Describe the results you received:**

```
*   Trying /var/run/docker.sock...
* Connected to localhost (/var/run/docker.sock) port 80 (#0)
> POST /images/create?fromImage=127.0.0.1:5000/docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8 HTTP/1.1
> Host: localhost
> User-Agent: curl/7.52.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Api-Version: 1.26
< Content-Type: application/json
< Docker-Experimental: true
< Server: Docker/17.03.0-ce (linux)
< Date: Fri, 24 Mar 2017 22:51:03 GMT
< Transfer-Encoding: chunked
<
{"status":"Pulling repository 127.0.0.1:5000/docker/build_context"}
{"errorDetail":{"message":"Error: image docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8 not found"},"error":"Error: image docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8 not found"}
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
```

**Describe the results you expected:**

```
*   Trying /var/run/docker.sock...
* Connected to localhost (/var/run/docker.sock) port 80 (#0)
> POST /images/create?fromImage=127.0.0.1:5000/docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8 HTTP/1.1
> Host: localhost
> User-Agent: curl/7.52.1
> Accept: */*
>
< HTTP/1.1 404 Not Found
< Api-Version: 1.26
< Content-Type: application/json
< Docker-Experimental: true
< Server: Docker/17.03.0-ce (linux)
< Date: Fri, 24 Mar 2017 22:51:03 GMT
< Transfer-Encoding: chunked
<
{"message":"repository docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8 not found: does not exist or no pull access"}
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:34:06 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:34:06 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 6
 Running: 6
 Paused: 0
 Stopped: 0
Images: 23
Server Version: 17.03.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: active
 NodeID: oilmywug34uqn4jgm3rm32ibu
 Is Manager: true
 ClusterID: yr9hwk7g22nuk2dlq06nr4str
 Managers: 3
 Nodes: 8
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.31.24.104
 Manager Addresses:
  172.31.24.104:2377
  172.31.4.144:2377
  172.31.8.205:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: N/A (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.11-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.785 GiB
Name: ip-172-31-24-104.us-west-1.compute.internal
ID: G6IK:SX62:6HOM:LC57:VON4:PPS3:TZG5:K3NJ:KGSB:F6UO:TMJV:JN2U
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 139
 Goroutines: 493
 System Time: 2017-03-24T22:54:37.94104216Z
 EventsListeners: 2
Registry: https://index.docker.io/v1/
Labels:
 os=linux
 region=us-west-1
 availability_zone=us-west-1c
 instance_type=t2.large
 node_type=manager
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
This is a Docker for AWS swarm

**Comment by aaronlehmann** [comment]
I believe this happens because there is progress output written to the HTTP client before we find out whether the image exists (`{"status":"Pulling repository 127.0.0.1:5000/docker/build_context"}`).

**Comment by luxifer** [comment]
I'm digging

**Comment by thaJeztah** [comment]
Looks like this was resolved indeed (I didn't search which PR fixed it); tried on Docker 20.10 (both on Linux and Docker Desktop for Mac)


Pulling from a local registry (`127.0.0.1:5000/docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8` as the original example)

```bash
curl --unix-socket /var/run/docker.sock -X POST http://localhost/images/create\?fromImage\=127.0.0.1:5000/docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8 -v
*   Trying /var/run/docker.sock:0...
* Connected to localhost (/Users/sebastiaan/.docker/run/docker.sock) port 80 (#0)
> POST /images/create?fromImage=127.0.0.1:5000/docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8 HTTP/1.1
> Host: localhost
> User-Agent: curl/7.79.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Api-Version: 1.41
< Content-Type: application/json
< Date: Fri, 04 Feb 2022 09:06:38 GMT
< Docker-Experimental: true
< Ostype: linux
< Server: Docker/20.10.12 (linux)
< Transfer-Encoding: chunked
<
{"message":"manifest for 127.0.0.1:5000/docker/build_context:f3910548-cc8a-4211-8320-1c4a3b17fc7c-8 not found: manifest unknown: manifest unknown"}
* Connection #0 to host localhost left intact
```

Pulling from Docker Hub (`docker.io/library/nosuchimage:latest`):

```bash
curl --unix-socket /var/run/docker.sock -X POST http://localhost/images/create\?fromImage\=docker.io/library/nosuchimage:latest -v
*   Trying /var/run/docker.sock:0...
* Connected to localhost (/Users/sebastiaan/.docker/run/docker.sock) port 80 (#0)
> POST /images/create?fromImage=docker.io/library/nosuchimage:latest HTTP/1.1
> Host: localhost
> User-Agent: curl/7.79.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Api-Version: 1.41
< Content-Type: application/json
< Date: Fri, 04 Feb 2022 09:09:37 GMT
< Docker-Experimental: true
< Ostype: linux
< Server: Docker/20.10.12 (linux)
< Transfer-Encoding: chunked
<
{"message":"pull access denied for nosuchimage, repository does not exist or may require 'docker login': denied: requested access to the resource is denied"}
* Connection #0 to host localhost left intact
```

**Comment by fermayo** [comment]
Thanks for checking and following up @thaJeztah! Always amazed by your work ethic.
